### PR TITLE
feat: add public key support

### PR DESCRIPTION
**Pull request by vasco-santos**
Adding public key methods for allowing full interop with `go-ipfs`

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -164,13 +207,34 @@ const sign = (privateKey, value, validityType, validity, callback) => {
...
 }
 
-// Create record data for being signed
-const ipnsEntryDataForSig = (value, validityType, eol) => {
+// Utility for getting the validity type code name of a validity
+const getValidityType = (validityType) => {
+  if (validityType.toString() === '0') {
+    return 'EOL'
+  } else {
+    log.error('unrecognized validity type')
```
Comment (alanshaw) [inline src/index.js]: Put the `validityType` in the log message for better debugging

File: src/index.js [hunk 2]
```diff
@@ -139,16 +182,16 @@ const getLocalKey = (key) => new Key(`/ipns/${rawStdEncoding(key)}`)
  * Get key for sharing the record in the routing mechanism.
  * Format: ${base32(/ipns/<HASH>)}, ${base32(/pk/<HASH>)}
  *
- * @param {Buffer} key peer identifier object.
+ * @param {Buffer} pid peer identifier object.
```
Comment (alanshaw) [inline src/index.js]: Not specifically an object, a buffer. Does this need an explanation of how you might extract this from a `PeerId` instance?

File: src/index.js [hunk 3]
```diff
@@ -100,15 +104,48 @@ const validate = (publicKey, entry, callback) => {
...
+      extractedPublicKey = extractPublicKeyFromId(peerIdResult)
+    } catch (err) {
+      const error = 'cannot get public key from the peer id'
+
+      log.error(error)
+      return callback(Object.assign(new Error(error), { code: ERRORS.ERR_PUBLICK_KEY_FROM_ID }))
+    }
+
+    if (extractedPublicKey) {
+      return callback(null, null)
```
[the remark this answers is not in the document]
Comment (vasco-santos) [reply src/index.js]: According to the docs:

If the `resultEntry` is also null, the `peer-id` allows to extract the public key from the `peer-id` and there is no need in extracting it.

I am using the same approach as `go-ipns`, and consequently returning null if the entry was not modified. 

Regarding the test, you are right and I will look into that!

File: src/index.js [hunk 4]
```diff
@@ -13,6 +16,7 @@ const ipnsEntryProto = require('./pb/ipns.proto')
 const { parseRFC3339 } = require('./utils')
 const ERRORS = require('./errors')
 
+const ID_MULTIHASH_CODE = 0
```
[the remark this answers is not in the document]
Comment (vasco-santos) [reply src/index.js]: Just created a PR https://github.com/multiformats/js-multihash/pull/57 for that 🙂

File: src/index.js [hunk 5]
```diff
@@ -120,7 +153,17 @@ const embedPublicKey = (publicKey, entry, callback) => {
  * @return {Void}
  */
 const extractPublicKey = (peerId, entry, callback) => {
-  callback(new Error('not implemented yet'))
+  if (entry.pubKey) {
+    try {
+      const pubKey = crypto.keys.unmarshalPublicKey(entry.pubKey)
+
+      return callback(null, pubKey)
```
Comment (alanshaw) [inline src/index.js]: Avoid callback in a `try` block - if the callback throws, it'll be caught and the callback will be called again - very hard to debug!

File: src/index.js [hunk 6]
```diff
@@ -100,15 +104,44 @@ const validate = (publicKey, entry, callback) => {
...
+ * send this as part of the record itself. For newer ed25519 keys, the public key
+ * can be embedded in the peerId.
  *
- * @param {Object} publicKey public key for validating the record.
+ * @param {Object} publicKey public key to embed.
  * @param {Object} entry ipns entry record.
  * @param {function(Error)} [callback]
  * @return {Void}
  */
 const embedPublicKey = (publicKey, entry, callback) => {
```
Comment (alanshaw) [inline src/index.js]: Check `publicKey`, `publicKey.bytes` and `entry` exist?

File: src/index.js [hunk 7]
```diff
@@ -100,15 +104,44 @@ const validate = (publicKey, entry, callback) => {
...
+      log.error(err)
+      return callback(Object.assign(new Error(err), { code: ERRORS.ERR_PUBLIC_KEY_FROM_ID }))
+    }
+
+    if (extractedPublicKey) {
+      return callback(null, null)
+    }
+
+    // If we failed to extract the public key from the peer ID, embed it in the record.
+    entry.pubKey = crypto.keys.marshalPublicKey(publicKey)
```
Comment (alanshaw) [inline src/index.js]: try/catch around this?

File: src/index.js [hunk 8]
```diff
@@ -120,7 +153,17 @@ const embedPublicKey = (publicKey, entry, callback) => {
  * @return {Void}
  */
 const extractPublicKey = (peerId, entry, callback) => {
```
Comment (alanshaw) [inline src/index.js]: Check `peerId` and `entry` exist

File: package.json [hunk 9]
```diff
@@ -49,8 +51,7 @@
     "dirty-chai": "^2.0.1",
     "ipfs": "^0.29.3",
     "ipfsd-ctl": "^0.36.0",
-    "libp2p-crypto": "^0.13.0",
-    "multihashes": "^0.4.13"
+    "libp2p-crypto": "^0.13.0"
```
Comment (alanshaw) [inline package.json]: This needs to move to `dependencies`

File: src/index.js (hunk 5 above)
Comment (vasco-santos) [reply src/index.js]: true!

**Pull request merged**